enhancements/storage/fsgroupchange-policy.md

```diff
@@ -0,0 +1,103 @@
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+We want to allow Openshift admins to change `FSGroupChangePolicy` defaulting policies for certain namespaces so as pods in the given namespace can start faster.
+
+## Motivation
+
+Currently it can take a long time for volume's permission to be changed to match pod's `fsGroup`. This is a well known problem
+in Kubernetes and Openshift. See - https://access.redhat.com/solutions/6221251 for more details.
+
+To solve this problem, we introduced `FSGroupChangePolicy` - https://github.com/kubernetes/kubernetes/blob/b5608aea94cfb54fea3a63e1d74235759d036c51/pkg/apis/core/types.go#L3977 , which
+can be configured in a `PodSpec` and be set to `OnRootMismatch`, which instructs kubelet to stop recursively changing permissions of a volume, if top level directory permissions match as expected by kubelet.
+
+This usually speeds up pod startup because kubelet does not have to recursively change permission of each file and directory in the volume. This mechanism relies on heuristic that gets applied when recursively changing permissions of a volume -  kubelet changes permissions of top level directory as a last step in the entire process, when it performs full recursive `chown` and `chmod` of the volume.
+
+Having said that, `FSGroupChangePolicy` still defaults to `Always`, which means kubelet will always recursively change permissions. The reason for defaulting to `Always` is - mostly backward compatibility and if users bring in volumes that were used outside kubernetes. In these cases `OnRootMismatch` may not be a safe default.
+
+But for most intent and purpose, it can be a reasonable default and hence this proposal allows admins to configure a namespace such that, all pods created in the namespace will default to `OnRootMismatch` policy if pod does not specify a policy of its own. If a pod specifies its own `FSGroupChangePolicy` that is still used as a default and will not be overridden.
+
+### Goals
+
+Allow Openshift admins to configure per-namespace default of `FSGroupChangePolicy`.
+
+## Non-Goals
+
+* Change Openshift default globally.
+
+## Proposal
+
+### Allow admins to opt-in to `FSGroupChangePolicy` via namespace annotation
+
+We propose an annotation `openshift.io/fsgroup-change-policy` if that is set to `OnRootMismatch` in a namespace, then Openshift's SCC hooks start defaulting all pods created in that namespace with `OnRootMismatch` `FSGroupChangePolicy`.
+
+```go
+func getPodFsGroupChangePolicy(ns *corev1.Namespace) *api.PodFSGroupChangePolicy {
+    fsGroupPolicy, ok := ns.Annotations[securityv1.OnRootMismatchFSGroupPolicy]
+    if !ok {
+        return nil
+    }
+    if fsGroupPolicy == "OnRootMismatch" {
+        onRootMismatchPolicy := api.FSGroupChangeOnRootMismatch
+        return &onRootMismatchPolicy
+    }
+    return nil
+}
+```
+
+### Set default `FSGroupChangePolicy`in pods during admission:
```

Comment:
Generally when I've seen guidance changes like this in the past, we've effected that change throughout operators in OpenShift as well. Is there any plan for forcing this value on OpenShift operators/operands?

Reply:
It could be useful for operators that use persistent volumes, such as  - alertmanager, prometheus and image-registry etc. Having said that, it is relatively easier to modify pod spec of OCP operators to have this field, but the idea is - a typical web developer may not be even aware of existance of this option, so having an OCP admin set this defaulting for developer namespaces will speed up their workloads even if workload author's aren't aware of the option.